pkg/runtime/handlerTable.go

```diff
@@ -0,0 +1,178 @@
+type HandlerEntry struct {
+	// Name of the handler
+	Name string
+
+	// Handler is the initialized handler object.
+	Handler adapter.Handler
+
+	// HandlerCreateError records error while creating the handler
+	HandlerCreateError error
+
+	// Instances is the global list of instances associated with this handler
+	Instances map[string]bool
+
+	// sha is used to verify and update the handlerEntry.
+	// sha must use handler.Params since any change to Params
+	// requires creation of a new handler instance.
+	// Some handlers need to be told about data they are about to receive.
+	// For example prometheus adapter initialized objects based on this configuration.
+	// If new configInstances are associated with such a handler, it also necessitates
+	// recreation of the handler.
+	sha [sha1.Size]byte
+
+	// If this is set, the handler is closed.
+	notInUse bool
+}
+
+func newHandlerTable(instanceConfig map[string]*cpb.Instance, handlerConfig map[string]*cpb.Handler,
+	buildHandler buildHandlerFn) *handlerTable {
+	return &handlerTable{
+		instanceConfig: instanceConfig,
+		handlerConfig:  handlerConfig,
+		buildHandler:   buildHandler,
+	}
+}
+
+// handlerTable initializes and maintains handlers.
+// It can efficiently initialize the next version of handlers based
+// on the previous handler table.
+type handlerTable struct {
+	instanceConfig map[string]*cpb.Instance
+	handlerConfig  map[string]*cpb.Handler
+
+	// Used to build handler
+	buildHandler buildHandlerFn
+
+	// table that maintains handler state
+	table map[string]*HandlerEntry
+}
+
+// Associate an instance with a handler
+func (t *handlerTable) Associate(handleName string, instanceName string) {
+	h := t.table[handleName]
+	if h == nil {
+		h = &HandlerEntry{
+			Name:      handleName,
+			Instances: make(map[string]bool),
+		}
+		t.table[handleName] = h
+	}
+	h.Instances[instanceName] = true
+}
+
+// Initialize the handler table based on configuration and the old handler table.
+// If handler config has not changed, connections from the old handler table are re-used.
+// This method does not return an error, it records errors in the handlerEntry.
+func (t *handlerTable) Initialize(oldTable map[string]*HandlerEntry) {
+	t.computeSha()
+	// run diff with the old handlerTable
+	for oh, ohe := range oldTable {
+		var he *HandlerEntry
+		if he = t.table[oh]; he == nil {
+			// old handler is not needed any more
+			ohe.notInUse = true
+			if glog.V(3) {
+				glog.Infof("handler: %s will be removed", oh)
+			}
+			continue
+		}
+
+		if he.sha != ohe.sha {
+			ohe.notInUse = true
+			if glog.V(3) {
+				glog.Infof("handler: %s will be replaced", oh)
+			}
+			continue
+		}
+		// shas match reuse
+		he.Handler = ohe.Handler
+	}
+
+	// initialize handlers that were not previously covered.
+	for _, he := range t.table {
+		// this was already initialized.
+		if he.Handler != nil {
+			continue
+		}
+		// create a new handler
+		// handler error is marked inside the entry.
+		t.initHandler(he)
```

Comment:
We cannot do it there. The purge loop runs on the  *oldTable* and copies over handler that can be re-used.  
The 2nd loop runs on the *newTable*